### PR TITLE
feat(cli): add price max precision checks for orders

### PR DESCRIPTION
**Pull request by engwarrior**
https://github.com/ExchangeUnion/xud/issues/1343. We decided to go with 10 ** 12 max precision. The number is easily configurable any time. Actually was quite a challenge to setup xud with lnd, btcd, ltcd and still shows some swapClient when executing order, but precision checks work fine.
Screenshot 2020-02-04 at 15 12 24

### CODE REVIEW

**Comment by engwarrior** [comment]
CI build failed but it's not related to tests or lint or built: its setting up simulation which fails on 2020/02/04 12:49:29 cannot start xud network: context deadline exceeded

File: lib/cli/placeorder.ts [hunk 1]
```diff
@@ -1,20 +1,21 @@
...
 import { Order, OrderSide, PlaceOrderEvent, PlaceOrderRequest, PlaceOrderResponse, SwapFailure, SwapSuccess } from '../proto/xudrpc_pb';
 import { callback, loadXudClient } from './command';
-import { coinsToSats, satsToCoinsStr } from './utils';
+import { checkPrecision, coinsToSats, satsToCoinsStr } from './utils';
 
 export const placeOrderBuilder = (argv: Argv, side: OrderSide) => {
   const command = side === OrderSide.BUY ? 'buy' : 'sell';
   argv.option('quantity', {
-    type: 'number',
+    type: 'string',
```
Comment (sangaman) [inline lib/cli/placeorder.ts]: I think it's better to keep this as a number, since that's what it is, and make the string conversion (if necessary) later in the code.

File: lib/cli/utils.ts [hunk 2]
```diff
@@ -4,6 +4,7 @@ import os from 'os';
 import path from 'path';
 
 const SATOSHIS_PER_COIN = 10 ** 8;
+const MAX_PRECISION = 12;
```
Comment (sangaman) [inline lib/cli/utils.ts]: Maybe call this `MAX_PRICE_DECIMAL_PLACES`. "Decimal place" is more accurate in this context I believe because "precision" can refer to the number of significant digits either to the left or right of the decimal point.

File: lib/cli/utils.ts [hunk 3]
```diff
@@ -33,6 +34,18 @@ export const coinsToSats = (coinsQuantity: number) => {
   return Math.round(coinsQuantity * SATOSHIS_PER_COIN);
 };
 
+/** Checks order price precision is not more than MAX_PRECISION. */
+export const checkPrecision = (price: string) => {
```
Comment (sangaman) [inline lib/cli/utils.ts]: `checkDecimalPlaces` maybe for same reasoning as above?

File: lib/cli/placeorder.ts [hunk 4]
```diff
@@ -1,20 +1,21 @@
...
+    describe: 'the quantity for limit/market orders, max 12 digits on the right after a decimal point',
   })
   .option('pair_id', {
     type: 'string',
     describe: 'the trading pair ticker id for the order',
   })
   .option('price', {
     type: 'string',
-    describe: 'the price for limit orders, or "mkt"/"market" for market orders',
+    describe: 'the price for limit orders, or "mkt"/"market" for market orders, max 12 digits on the right after a decimal point',
```
Comment (sangaman) [inline lib/cli/placeorder.ts]: "max 12 decimal places" ? just a bit shorter that way and there's not much room in terminals

File: lib/cli/placeorder.ts (hunk 1 above)
Comment (engwarrior) [reply lib/cli/placeorder.ts]: It's not that simple unfortunately, because if the user sets the quantity to 1e-24 for example it will be converted with .toString() to "1e-24".

**Comment by engwarrior** [comment]
I will add to RPC too, its very fast. But you probably mean gRPC right? I see only gRPC https://github.com/ExchangeUnion/xud/blob/master/lib/grpc/GrpcService.ts#L15

File: lib/cli/placeorder.ts (hunk 1 above)
Comment (engwarrior) [reply lib/cli/placeorder.ts]: in case if we want to use toFixed(decimals), we will have to specify large number of decimals like 100 for potential user case when people will specify 10 ** 100 precision and check all digits after decimal point if they are not 0.

**Comment by engwarrior** [comment]
RPC logic is very trivial after testing new method implementation. I changed that to fit number rather than string as requested and being cleaner.

File: package.json [hunk 5]
```diff
@@ -129,6 +129,7 @@
   ],
   "dependencies": {
     "@exchangeunion/grpc-dynamic-gateway": "^0.3.10",
+    "bignumber.js": "^9.0.0",
```
Comment (erkarl) [inline package.json]: I'm not sure about adding this dependency. Is this really necessary?

Comment (engwarrior) [reply package.json]: I didn't find more elegant and correct way without checking out corner cases to check if Number decimal places are at most 12.

**Comment by engwarrior** [comment]
@erkarl @sangaman @kilrau are going to review this guys?

File: lib/service/Service.ts [hunk 6]
```diff
@@ -58,6 +59,8 @@ const argChecks = {
   POSITIVE_AMOUNT: ({ amount }: { amount: number }) => { if (amount <= 0) throw errors.INVALID_ARGUMENT('amount must be greater than 0'); },
   POSITIVE_QUANTITY: ({ quantity }: { quantity: number }) => { if (quantity <= 0) throw errors.INVALID_ARGUMENT('quantity must be greater than 0'); },
   PRICE_NON_NEGATIVE: ({ price }: { price: number }) => { if (price < 0) throw errors.INVALID_ARGUMENT('price cannot be negative'); },
+  QUANTITY_DECIMAL_PLACES: ({ quantity }: { quantity: number }) => { if (!checkDecimalPlaces(quantity)) throw errors.INVALID_ARGUMENT('quantity mush have max 12 decimal places'); },
+  PRICE_DECIMAL_PLACES: ({ price }: { price: number }) => { if (!checkDecimalPlaces(price)) throw errors.INVALID_ARGUMENT('price mush have max 12 decimal places'); },
```
Comment (michael1011) [inline lib/service/Service.ts]:
```suggestion
  PRICE_DECIMAL_PLACES: ({ price }: { price: number }) => { if (!checkDecimalPlaces(price)) throw errors.INVALID_ARGUMENT('price must have less than 13 decimals); },
```

Maybe reword the error so that it is more clear?

File: lib/service/Service.ts [hunk 7]
```diff
@@ -58,6 +59,8 @@ const argChecks = {
   POSITIVE_AMOUNT: ({ amount }: { amount: number }) => { if (amount <= 0) throw errors.INVALID_ARGUMENT('amount must be greater than 0'); },
   POSITIVE_QUANTITY: ({ quantity }: { quantity: number }) => { if (quantity <= 0) throw errors.INVALID_ARGUMENT('quantity must be greater than 0'); },
   PRICE_NON_NEGATIVE: ({ price }: { price: number }) => { if (price < 0) throw errors.INVALID_ARGUMENT('price cannot be negative'); },
+  QUANTITY_DECIMAL_PLACES: ({ quantity }: { quantity: number }) => { if (!checkDecimalPlaces(quantity)) throw errors.INVALID_ARGUMENT('quantity mush have max 12 decimal places'); },
```
Comment (michael1011) [inline lib/service/Service.ts]: Same here

File: lib/cli/placeorder.ts [hunk 8]
```diff
@@ -62,6 +63,11 @@ export const placeOrderHandler = (argv: Arguments<any>, side: OrderSide) => {
     return;
   }
 
+  if (!checkDecimalPlaces(numericPrice) || !checkDecimalPlaces(argv.quantity)) {
+    console.log('price and quantity must have maximum 12 decimal places');
```
Comment (michael1011) [inline lib/cli/placeorder.ts]: Same here

File: lib/cli/placeorder.ts (hunk 1 above)
Comment (sangaman) [reply lib/cli/placeorder.ts]: You can do `toFixed` with 12 decimal places, then compare the price to the fixed string.

```typescript
const fixedPrice = price.toFixed(12);
if (price > fixedPrice) // we have numbers after the first 12 decimal places, invalid argument
```

File: package.json (hunk 5 above)
Comment (sangaman) [reply package.json]: I agree that this is not necessary, see my comment above about a simpler way to check if we have more than 12 decimal places.

File: lib/cli/placeorder.ts (hunk 1 above)
Comment (engwarrior) [reply lib/cli/placeorder.ts]: @sangaman I hope JS compares doubles correctly.

Comment (sangaman) [reply lib/cli/placeorder.ts]: What is your concern? All numbers are doubles in JS.

Comment (engwarrior) [reply lib/cli/placeorder.ts]: @sangaman its better to avoid comparison of doubles with high precision

**Comment by engwarrior** [comment]
Removed bignumber.js

**Comment by erkarl** [comment]
Looks like this PR now deletes `npm-shrinkwrap.json`.

File: lib/cli/placeorder.ts [hunk 9]
```diff
@@ -1,20 +1,21 @@
...
 import { Order, OrderSide, PlaceOrderEvent, PlaceOrderRequest, PlaceOrderResponse, SwapFailure, SwapSuccess } from '../proto/xudrpc_pb';
 import { callback, loadXudClient } from './command';
-import { coinsToSats, satsToCoinsStr } from './utils';
+import { checkDecimals, coinsToSats, satsToCoinsStr } from './utils';
 
 export const placeOrderBuilder = (argv: Argv, side: OrderSide) => {
   const command = side === OrderSide.BUY ? 'buy' : 'sell';
   argv.option('quantity', {
     type: 'number',
+    describe: 'the quantity for limit/market orders, less than 13 decimals',
```
Comment (sangaman) [inline lib/cli/placeorder.ts]: We aren't applying this 12 max decimal places limit to quantity, that is already limited to 8 (satoshis). We can leave this alone.

File: lib/cli/placeorder.ts [hunk 10]
```diff
@@ -1,20 +1,21 @@
...
   })
   .option('pair_id', {
     type: 'string',
     describe: 'the trading pair ticker id for the order',
   })
   .option('price', {
     type: 'string',
-    describe: 'the price for limit orders, or "mkt"/"market" for market orders',
+    describe: 'the price for limit orders, or "mkt"/"market" for market orders, less than 13 decimals',
```
Comment (sangaman) [inline lib/cli/placeorder.ts]: I personally think "max 12 decimal places" is more clear - although I also think this description can also be left alone as in most cases people aren't going to be trying to enter prices with that many decimal places anyway. Maybe better not to confuse anyone, if somebody tries a really weird price then they will get a more descriptive error message.

File: lib/cli/placeorder.ts [hunk 11]
```diff
@@ -62,6 +63,11 @@ export const placeOrderHandler = (argv: Arguments<any>, side: OrderSide) => {
     return;
   }
 
+  if (checkDecimals(numericPrice) || checkDecimals(argv.quantity)) {
+    console.log('price and quantity must have less than 13 decimals');
```
Comment (sangaman) [inline lib/cli/placeorder.ts]: Again there shouldn't be a check on quantity here.

File: lib/cli/utils.ts [hunk 12]
```diff
@@ -4,6 +4,7 @@ import os from 'os';
 import path from 'path';
 
 const SATOSHIS_PER_COIN = 10 ** 8;
+const MAX_DECIMALS = 12;
```
Comment (sangaman) [inline lib/cli/utils.ts]: I'd prefer what I suggested originally so it is clear what this is for, `MAX_PRICE_DECIMAL_PLACES`

File: lib/service/Service.ts [hunk 13]
```diff
@@ -58,6 +59,8 @@ const argChecks = {
   POSITIVE_AMOUNT: ({ amount }: { amount: number }) => { if (amount <= 0) throw errors.INVALID_ARGUMENT('amount must be greater than 0'); },
   POSITIVE_QUANTITY: ({ quantity }: { quantity: number }) => { if (quantity <= 0) throw errors.INVALID_ARGUMENT('quantity must be greater than 0'); },
   PRICE_NON_NEGATIVE: ({ price }: { price: number }) => { if (price < 0) throw errors.INVALID_ARGUMENT('price cannot be negative'); },
+  QUANTITY_CORRECT_PRECISION: ({ quantity }: { quantity: number }) => { if (checkDecimals(quantity)) throw errors.INVALID_ARGUMENT('quantity must have less than 13 decimal'); },
```
Comment (sangaman) [inline lib/service/Service.ts]: We don't need this for quantity. Also on the grpc layer quantity will always be an integer, as it represents whole numbers of satoshis, so it would never fail this check.

File: lib/cli/utils.ts [hunk 14]
```diff
@@ -33,6 +34,12 @@ export const coinsToSats = (coinsQuantity: number) => {
   return Math.round(coinsQuantity * SATOSHIS_PER_COIN);
 };
 
+/** Checks max decimals. */
```
Comment (sangaman) [inline lib/cli/utils.ts]: It's not clear what this does based on the description, it looks like earlier this returned `true` when the check passed, but now it returns `true` when the check fails. So better to specify here that the function returns true when the decimal places are exceeded.

File: lib/service/Service.ts [hunk 15]
```diff
@@ -58,6 +59,8 @@ const argChecks = {
...
+  PRICE_CORRECT_PRECISION: ({ price }: { price: number }) => { if (checkDecimals(price)) throw errors.INVALID_ARGUMENT('price must have less than 13 decimal'); },
```
Comment (sangaman) [inline lib/service/Service.ts]: Let's go with `PRICE_MAX_DECIMAL_PLACES` for the check name and `price cannot have more than 12 decimal places` for the message, if you don't mind.

**Comment by flicker-harmony** [comment]
travis fails because of private key https://travis-ci.org/ExchangeUnion/xud/builds/646934061#L854 locally tests are fine

File: lib/cli/placeorder.ts [hunk 16]
```diff
@@ -2,11 +2,13 @@ import { Arguments, Argv } from 'yargs';
 import { Order, OrderSide, PlaceOrderEvent, PlaceOrderRequest, PlaceOrderResponse, SwapFailure, SwapSuccess } from '../proto/xudrpc_pb';
 import { callback, loadXudClient } from './command';
 import { coinsToSats, satsToCoinsStr } from './utils';
+import { checkDecimalPlaces } from '../utils/utils';
 
 export const placeOrderBuilder = (argv: Argv, side: OrderSide) => {
   const command = side === OrderSide.BUY ? 'buy' : 'sell';
   argv.option('quantity', {
     type: 'number',
+    describe: 'the quantity for limit/market orders',
```
Comment (sangaman) [inline lib/cli/placeorder.ts]: We have some order besides limit and market, so I'd either leave this line out (not needed for this PR anyhow) or just say something more generic like "the quantity to trade".

File: lib/cli/placeorder.ts [hunk 17]
```diff
@@ -62,6 +64,11 @@ export const placeOrderHandler = (argv: Arguments<any>, side: OrderSide) => {
     return;
   }
 
+  if (checkDecimalPlaces(numericPrice)) {
+    console.log('price cannot have more than 12 decimal places');
```
Comment (sangaman) [inline lib/cli/placeorder.ts]: We should log this to `console.error` and also set the status code to a non-zero value to indicate the call has failed with `process.exitCode = 1`.

File: lib/cli/placeorder.ts (hunk 16 above)
Comment (flicker-harmony) [reply lib/cli/placeorder.ts]: Done

File: lib/cli/placeorder.ts (hunk 17 above)
Comment (flicker-harmony) [reply lib/cli/placeorder.ts]: Done

**Comment by erkarl** [comment]
Looks good :+1:.

As a final nit let's clean up the commit message and PR title to exclude `and quantity`.

**Comment by engwarrior** [comment]
Fails integration on travis...

**Comment by erkarl** [comment]
> Fails integration on travis...

We're experiencing major flakiness in our simulation tests - I restarted your build. Things should improve a lot once https://github.com/ExchangeUnion/xud/pull/1357 gets merged.

**Comment by engwarrior** [comment]
Still

**Comment by engwarrior** [comment]
Doesn't help. Is something broken? I don't understand test that is failing. And it passed before.

**Comment by kilrau** [comment]
Restarted again. It's annoying, I know but improvement is on the way (https://github.com/ExchangeUnion/xud/pull/1357).

**Comment by engwarrior** [comment]
Maybe my PR broke integration test? But I can't check it locally and it was okay before.

**Comment by erkarl** [comment]
> Maybe my PR broke integration test? But I can't check it locally and it was okay before.

All good on your side. This will need to get merged before the test suite will pass: https://github.com/ExchangeUnion/xud/pull/1386

**Comment by engwarrior** [comment]
Will you merge tests fix PR and merge mine today?

**Comment by kilrau** [comment]
Just merged the fix and restarted tests. Thanks for your patience @engwarrior

**Comment by flicker-harmony** [comment]
@kilrau merge.

**Comment by kilrau** [comment]
> @kilrau merge.

Aye, sir ;)
